x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DatafeedContextProvider.java

```diff
@@ -38,27 +34,19 @@ public DatafeedContextProvider(
     public void buildDatafeedContext(String datafeedId, ActionListener<DatafeedContext> listener) {
         DatafeedContext.Builder context = DatafeedContext.builder();
 
-        Consumer<DatafeedTimingStats> timingStatsListener = timingStats -> {
-            context.setTimingStats(timingStats);
-            listener.onResponse(context.build());
-        };
-
-        ActionListener<RestartTimeInfo> restartTimeInfoListener = ActionListener.wrap(restartTimeInfo -> {
-            context.setRestartTimeInfo(restartTimeInfo);
-            resultsProvider.datafeedTimingStats(context.getJob().getId(), timingStatsListener, listener::onFailure);
-        }, listener::onFailure);
-
-        ActionListener<Job.Builder> jobConfigListener = ActionListener.wrap(jobBuilder -> {
-            context.setJob(jobBuilder.build());
-            resultsProvider.getRestartTimeInfo(jobBuilder.getId(), restartTimeInfoListener);
-        }, listener::onFailure);
-
-        ActionListener<DatafeedConfig.Builder> datafeedListener = ActionListener.wrap(datafeedConfigBuilder -> {
+        datafeedConfigProvider.getDatafeedConfig(datafeedId, null, listener.delegateFailureAndWrap((delegate1, datafeedConfigBuilder) -> {
             DatafeedConfig datafeedConfig = datafeedConfigBuilder.build();
             context.setDatafeedConfig(datafeedConfig);
-            jobConfigProvider.getJob(datafeedConfig.getJobId(), null, jobConfigListener);
-        }, listener::onFailure);
-
-        datafeedConfigProvider.getDatafeedConfig(datafeedId, null, datafeedListener);
+            jobConfigProvider.getJob(datafeedConfig.getJobId(), null, delegate1.delegateFailureAndWrap((delegate2, jobBuilder) -> {
+                context.setJob(jobBuilder.build());
+                resultsProvider.getRestartTimeInfo(jobBuilder.getId(), delegate2.delegateFailureAndWrap((delegate3, restartTimeInfo) -> {
+                    context.setRestartTimeInfo(restartTimeInfo);
```

Comment:
I don't think it's great for readability to nest these things so deeply (or even really at all). Could we go back to declaring the listeners up-front? A `SubscribableListener#andThen` chain would also be ok since we're not on a hot path here.

Reply:
> A SubscribableListener#andThen chain would also be ok since we're not on a hot path here.

I kinda started not liking that one as much any more for "transporty" things. It's very hard to reason about ref-counts with it. 

I'm with you that this code is far from beautiful but in terms of reasoning through that nothing leaks this is the easiest pattern IMO (that's kinda what I'm doing this for to begin with, the performance gains are trivial here obviously :)).
E.g. Declaring a bunch of listeners that all have the same listener as their failure callback makes it remarkably hard to be confident that there's no double release etc. (we even had 2 separate incidents of double releasing in precisely this kind of code ... EQL).

Also, at least here I can see what the callback is for what each call. Sure, if I want to know what is called by `getRestartTimeInfo` I need to also read over the callback inside `datafeedTimingStats` but on the flip-side I see exactly what sequence of events lead to each of these async calls and I know with 100% certainty that I won't leak stuff even on unexpected exceptions in the callback bodies. I don't have that with separately declared listeners?

Reply:
> I kinda started not liking that one as much any more for "transporty" things. It's very hard to reason about ref-counts with it.

Yeah but this seems fixable, it forces a pretty simple ownership pattern at least. And doesn't apply here AFAICT?

> E.g. Declaring a bunch of listeners that all have the same listener as their failure callback makes it remarkably hard to be confident that there's no double release etc.

Sure, I'm with you there, but that's because today they're declared with a raw `wrap()` call. We can declare each one as a `delegateFailureAndWrap()` around the previous one, I'd just rather we kept them as named symbols rather than inlining them all like this.

Reply:
Actually, now that I look at it, with the nested structure, we can follow up with a change to delete the `DatafeedContext.Builder` object altogether and just make the constructor package-private (or turn it into a `record`):

```java
    public void buildDatafeedContext(String datafeedId, ActionListener<DatafeedContext> listener) {
        datafeedConfigProvider.getDatafeedConfig(datafeedId, null, listener.delegateFailureAndWrap((delegate1, datafeedConfigBuilder) -> {
            DatafeedConfig datafeedConfig = datafeedConfigBuilder.build();
            jobConfigProvider.getJob(datafeedConfig.getJobId(), null, delegate1.delegateFailureAndWrap((delegate2, jobBuilder) -> {
                resultsProvider.getRestartTimeInfo(jobBuilder.getId(), delegate2.delegateFailureAndWrap((delegate3, restartTimeInfo) -> {
                    resultsProvider.datafeedTimingStats(jobBuilder.getId(), timingStats -> {
                        delegate3.onResponse(new DatafeedContext(datafeedConfig, jobBuilder.build(), restartTimeInfo, timingStats));
                    }, delegate3::onFailure);
                }));
            }));
        }));
    }
```

Reply:
Alright given the fact that I cleaned up one set of redundant variables via unnamed nesting and we have room to do the same here I'm going to go with. the existing version for now then, hope that's ok with everyone :)